File: qadevOOo/tests/java/ifc/sheet/_XCalculatable.java
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This file incorporates work covered by the following license notice:
 *
 *   Licensed to the Apache Software Foundation (ASF) under one or more
 *   contributor license agreements. See the NOTICE file distributed
 *   with this work for additional information regarding copyright
 *   ownership. The ASF licenses this file to you under the Apache
 *   License, Version 2.0 (the "License"); you may not use this file
 *   except in compliance with the License. You may obtain a copy of
 *   the License at http://www.apache.org/licenses/LICENSE-2.0 .
 */
package ifc.sheet;

import com.sun.star.sheet.XCalculatable;
import com.sun.star.table.XCell;
import lib.MultiMethodTest;
import lib.Status;
import lib.StatusException;

/**
 *
 */
public class _XCalculatable extends MultiMethodTest {
    public XCalculatable oObj = null;
    private boolean bIsAutomaticCalculationEnabled = false;
    private XCell[] xCells = null;

    /**
     * Get object relation: four cells with values and formulas.
     * @see mod._sc.ScModelObj
     */
    @Override
    public void before() {
        xCells = (XCell[])tEnv.getObjRelation("XCalculatable.Cells");
        if (xCells == null || xCells.length != 3)
            throw new StatusException(Status.failed("Couldn't find correct object relation 'XCalculatable.Cells'"));

    }

    /**
     * Restore begin setting
     */
    @Override
    public void after() {
        // reset to begin value
        oObj.enableAutomaticCalculation(bIsAutomaticCalculationEnabled);
    }


    public void _calculate() {
        requiredMethod("isAutomaticCalculationEnabled()");
        boolean result = true;
        double ergValue1 = xCells[2].getValue();
        double sourceValue1 = xCells[0].getValue();
        xCells[0].setValue(sourceValue1 +1);
        double ergValue2 = xCells[2].getValue();
        result &= ergValue1 == ergValue2;
        oObj.calculate();
        ergValue2 = xCells[2].getValue();
        result &= ergValue1 != ergValue2;
        tRes.tested("calculate()", result);
    }

    public void _calculateAll() {
        requiredMethod("isAutomaticCalculationEnabled()");
        boolean result = true;
        double ergValue1 = xCells[2].getValue();
        double sourceValue1 = xCells[0].getValue();
        xCells[0].setValue(sourceValue1 +1);
        double ergValue2 = xCells[2].getValue();
        result &= ergValue1 == ergValue2;
        oObj.calculateAll();
        ergValue2 = xCells[2].getValue();
        result &= ergValue1 != ergValue2;
        oObj.calculateAll();
        tRes.tested("calculateAll()", result);
    }

    public void _enableAutomaticCalculation() {
        bIsAutomaticCalculationEnabled = oObj.isAutomaticCalculationEnabled();
        oObj.enableAutomaticCalculation(!bIsAutomaticCalculationEnabled);
        tRes.tested("enableAutomaticCalculation()", true);
    }

    public void _isAutomaticCalculationEnabled() {
        requiredMethod("enableAutomaticCalculation()");
        boolean result = oObj.isAutomaticCalculationEnabled();
        oObj.enableAutomaticCalculation(false);
        tRes.tested("isAutomaticCalculationEnabled()", result != bIsAutomaticCalculationEnabled);
    }

}

File: test/source/sheet/xcalculatable.cxx
/* -*- Mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4; fill-column: 100 -*- */
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

#include <test/sheet/xcalculatable.hxx>

#include <com/sun/star/sheet/XCalculatable.hpp>
#include <com/sun/star/table/XCell.hpp>

#include <com/sun/star/uno/Reference.hxx>
#include <com/sun/star/uno/Sequence.hxx>

#include <cppunit/TestAssert.h>

using namespace com::sun::star;
using namespace com::sun::star::uno;

namespace apitest
{
void XCalculatable::testEnableAutomaticCaclulation()
{
    uno::Reference<sheet::XCalculatable> xCalculatable(init(), UNO_QUERY_THROW);

    bool bIsAutomaticCalculationEnabled = xCalculatable->isAutomaticCalculationEnabled();
    xCalculatable->enableAutomaticCalculation(!bIsAutomaticCalculationEnabled);
    bool bIsAutomaticCalculationEnabledNew = xCalculatable->isAutomaticCalculationEnabled();
    CPPUNIT_ASSERT_MESSAGE("Unable to set enableAutomaticCalculation()",
                           bIsAutomaticCalculationEnabled != bIsAutomaticCalculationEnabledNew);
}

void XCalculatable::testCalculate()
{
    uno::Reference<sheet::XCalculatable> xCalculatable(init(), UNO_QUERY_THROW);
    xCalculatable->enableAutomaticCalculation(false);

    uno::Sequence<uno::Reference<table::XCell>> xCells = getXCells();

    double dProduct = xCells[2]->getValue();
    double dFactor1 = xCells[0]->getValue();
    xCells[0]->setValue(dFactor1 + 1.0);
    double dProduct2 = xCells[2]->getValue();
    CPPUNIT_ASSERT_DOUBLES_EQUAL_MESSAGE("Calculated products are equal", dProduct, dProduct2, 0.5);

    xCalculatable->calculate();
    dProduct2 = xCells[2]->getValue();
    CPPUNIT_ASSERT_MESSAGE("Calculated products are not equal", dProduct != dProduct2);
}

void XCalculatable::testCalculateAll()
{
    uno::Reference<sheet::XCalculatable> xCalculatable(init(), UNO_QUERY_THROW);
    xCalculatable->enableAutomaticCalculation(false);

    uno::Sequence<uno::Reference<table::XCell>> xCells = getXCells();

    double dProduct = xCells[2]->getValue();
    double dFactor1 = xCells[0]->getValue();
    xCells[0]->setValue(dFactor1 + 1.0);
    double dProduct2 = xCells[2]->getValue();
    CPPUNIT_ASSERT_DOUBLES_EQUAL_MESSAGE("Calculated products are equal", dProduct, dProduct2, 0.5);

    xCalculatable->calculateAll();
    dProduct2 = xCells[2]->getValue();
    CPPUNIT_ASSERT_MESSAGE("Calculated products are not equal", dProduct != dProduct2);
    xCalculatable->calculateAll();
}
} // namespace apitest

/* vim:set shiftwidth=4 softtabstop=4 expandtab cinoptions=b1,g0,N-s cinkeys+=0=break: */
